host/http.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/julienschmidt/httprouter"
+	"github.com/flynn/flynn/host/types"
+	"github.com/flynn/flynn/pkg/httphelper"
+	"github.com/flynn/flynn/pkg/shutdown"
+	"github.com/flynn/flynn/pkg/sse"
+)
+
+type Host struct {
+	state   *State
+	backend Backend
+}
+
+func (h *Host) StopJob(id string) error {
+	job := h.state.GetJob(id)
+	if job == nil {
+		return errors.New("host: unknown job")
+	}
+	switch job.Status {
+	case host.StatusStarting:
+		h.state.SetForceStop(id)
+		return nil
+	case host.StatusRunning:
+		return h.backend.Stop(id)
+	default:
+		return errors.New("host: job is already stopped")
+	}
+}
+
+func (h *Host) streamEvents(id string, w http.ResponseWriter) error {
+	ch := h.state.AddListener(id)
+	go func() {
+		<-w.(http.CloseNotifier).CloseNotify()
+		h.state.RemoveListener(id, ch)
+	}()
+	enc := json.NewEncoder(sse.NewSSEWriter(w))
+	w.Header().Set("Content-Type", "text/event-stream; charset=utf-8")
+	w.WriteHeader(200)
+	w.(http.Flusher).Flush()
+	for {
+		select {
+		case data := <-ch:
+			if err := enc.Encode(data); err != nil {
+				return err
+			}
+			w.(http.Flusher).Flush()
+		case <-time.After(10 * time.Second):
+			// if the job still doesn't exist after a reasonable timeout, then
+			// break out of this stream (closing the connection will be handled outside)
+			if id != "all" && h.state.GetJob(id) == nil {
```

Comment:
Hm.  Yeah, we could retract this change.  I haven't seen it do any harm yet, and I'm not super pleased with the alternative behavior (hang), but yeah, I haven't checked for that kind of impacts.